src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -364,14 +365,15 @@ double ShenandoahAllocationRate::sample(size_t allocated) {
   double now = os::elapsedTime();
   double rate = 0.0;
   if (now - _last_sample_time > _interval_sec) {
-    if (allocated >= _last_sample_value) {
-      rate = instantaneous_rate(now, allocated);
-      _rate.add(rate);
-      _rate_avg.add(_rate.avg());
-    }
-
+    // _previous_gc_allocated will be non-zero only on the first sample added since the start of the current GC.  In the
+    // event that no data was sampled during the preceding cycle, _previous_gc_allocated will represent the accumalation
+    // of multiple GCs.
+    rate = instantaneous_rate(now, _previous_gc_allocated + allocated);
+    _rate.add(rate);
+    _rate_avg.add(_rate.avg());
     _last_sample_time = now;
```

Comment:
Note: in original implementation, the first sample() collected following the start of GC typically finds that allocated < _last_sample_value, so we ignore all data associated with that time gap.  With Shenandoah, the next sample happens the first time we ask should_start_gc() following the completion of the current GC.  With GenShen, the next sample happens after marking, when we calculate the allocation runway in order to size our reserves.

Reply:
It is possible that the first sample() collected() following the start of GC finds allocated > _last_sample value.  If this happens, the above computes a bogus result.  It thinks we've only allocated (new_sample_value - _last_sample_value) bytes in total time (now - _last_sample_time).  In truth, we've allocated new_sample_value plus currently_unknown_allocations following _last_sample_time but before the start of GC.